slick/src/sphinx/code/Connection.scala

```diff
@@ -103,6 +105,29 @@ object Connection extends App {
       val f: Future[Unit] = db.run(a)
       //#transaction
       Await.result(f, Duration.Inf)
+    };{
+      //#rollback
+      val insertCoffeesAction = coffees ++= Seq (
+        ("Cold_Drip", new SerialBlob(Array[Byte](101))),
+        ("Dutch_Coffee", new SerialBlob(Array[Byte](49)))
+      )
+
+      val countAction = coffees.length.result.map { coffeesCount: Int =>
+        if(coffeesCount < 5000) {
+          DBIO.failed(new Exception())
+        }
+      }
```

Comment:
I've been working a lot to figure out how best to convey that the rollback is a result of `DBIO.failed` and not a failed insert earlier in the transaction. I've come up with this sub-optimal solution. The idea is to show that coffee count is the same before and after, and by showing that the error is the same as what's in the `Exception`, to prove that's where it came from.

I couldn't chain `transactionally` with `asTry` directly for some reason, as it complained it didn't know what the `R` in `Try[R]` should be, but putting it separately seems to work, though it makes it harder to read.

I could also just show how to trigger the rollback, and have the reader trust that it just works, but that also seems sub-optimal.

I'd appreciate your thoughts here if you have a moment.

```scala
//#rollback
val countAction = coffees.length.result

val rollbackAction = (coffees ++= Seq(
  ("Cold_Drip", new SerialBlob(Array[Byte](101))),
  ("Dutch_Coffee", new SerialBlob(Array[Byte](49)))
)).flatMap { _ =>
  DBIO.failed(new Exception("Roll it back"))
}.transactionally

val errorHandleAction = rollbackAction.asTry.flatMap {
  case Failure(e: Throwable) => DBIO.successful(e.getMessage)
  case Success(_) => DBIO.successful("never reached")
}

val f = db.run(countAction zip errorHandleAction zip countAction).map {
  case ((initialCount, result), finalCount) =>
    // init: 5, final: 5, result: Roll it back
    println(s"init: ${initialCount}, final: ${finalCount}, result: ${result}")
    result
}

//#rollback
assert(Await.result(f, Duration.Inf) == "Roll it back")
```